## Cleaning And Exploring Ebay Car Sales Data

We will be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

|Columns   |  Description|
|:----------|:-------------|
|dateCrawled | When this ad was first crawled. All field-values are taken from this date|
|name | Name of the car|
|seller | Whether the seller is private or a dealer.|
|offerType | The type of listing|
|Price | The price on the ad to sell the car.|
|abtest | Whether the listing is included in an A/B test.
|vehicleType | The vehicle Type.|
|yearOfRegistration | The year in which which year the car was first registered.|
|gearbox | The transmission type.|
|powerPS | The power of the car in PS.|
|model | The car model name.|
|kilometer | How many kilometers the car has driven.|
|monthOfRegistration | The month in which which year the car was first registered.|
|fuelType | What type of fuel the car uses.|
|brand | The brand of the car.|
|notRepairedDamage | If the car has a damage which is not yet repaired.|
|dateCreated | The date on which the eBay listing was created.|
|nrOfPictures | The number of pictures in the ad.|
|postalCode | The postal code for the location of the vehicle.|
|lastSeenOnline | When the crawler saw this ad last online.|

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv(r'C:\Users\LENOVO\Desktop\projects\autos\data\autos.csv', encoding='Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

<b>Below we clean the column names

In [2]:
autos.rename({
    'dateCrawled':'date_crawled',
    'name':'name',
    'seller':'seller',
    'offerType':'offer_type',
    'price':'price',
    'abtest':'abtest',
    'vehicleType':'vehicle_type', 
    'yearOfRegistration':'registration_year',
    'gearbox':'gearbox', 
    'powerPS':'power_ps', 
    'model':'model',
    'odometer':'odometer', 
    'monthOfRegistration':'registration_month', 
    'fuelType':'fuel_type', 
    'brand':'brand',
    'notRepairedDamage':'unrepaired_damage', 
    'dateCreated':'ad_created', 
    'nrOfPictures':'no_of_pictures', 
    'postalCode':'postalCode',
    'lastSeen':'last_seen'
}, axis=1, inplace=True)

In [3]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postalCode            last_seen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Data Exploration and Cleaning  

In [4]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 16:37:21  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  no_of_pictures    postalCode            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [5]:
autos['price'].unique()
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

<b>We will clean the __price__ and __odometer__ column and convert to numeric dtype

In [6]:
autos['price'] = autos['price'].str.replace('$','').str.replace(',','').astype(int)
autos['odometer']  = autos['odometer'].str.replace(',','').str.replace('km','').astype(int)

#rename odometer column
autos.rename({'odometer':'odometer_kg'},axis=1,inplace=True)

In [7]:
#Check if what we did above is correct
autos['odometer_kg'].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000], dtype=int64)

<b> Exploring the Price and Odometer column

In [8]:
#price column
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Above we analyze the descriptive statistics and discovered that the minimum value in price is 0 which is awkward and maximum is 100,000,000,  we will explore te column further to ascertain what is wrong

In [9]:
autos['price'].value_counts().sort_index(ascending=True)


0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

In the above analysis we see there are  0 dollar prices meaning there are cars sold for  0 dollar which is unrealistic, also, some price suddenly jump from  350,000 to  999,990.

Therefore, we will categorise those unrealistic values as outiers which we are going to remove from our data

In [10]:
#Removing outliers
autos = autos[autos['price'].between(1,350000)]

#Check if what we did is correct
autos['price'].value_counts().sort_index(ascending=True)

1         156
2           3
3           1
5           2
8           1
9           1
10          7
11          2
12          3
13          2
14          1
15          2
17          3
18          1
20          4
25          5
29          1
30          7
35          1
40          6
45          4
47          1
49          4
50         49
55          2
59          1
60          9
65          5
66          1
70         10
         ... 
119500      1
119900      1
120000      2
128000      1
129000      1
130000      1
135000      1
137999      1
139997      1
145000      1
151990      1
155000      1
163500      1
163991      1
169000      1
169999      1
175000      1
180000      1
190000      1
194000      1
197000      1
198000      1
220000      1
250000      1
259000      1
265000      1
295000      1
299000      1
345000      1
350000      1
Name: price, Length: 2346, dtype: int64

In [11]:
#odometer_kg column
autos['odometer_kg'].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_kg, dtype: float64

In [12]:
autos['odometer_kg'].value_counts().sort_index(ascending=True)

5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_kg, dtype: int64

The analysis above shows that there is no outliers in __odometer_kg__ column, so we leave it as it is

### Exploring and Cleaning the date columns

In [13]:
#Exploring the date columns

autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In order for us to understand the date range, we first extract the first 10 character in the column, then use value_counts(normalize) to generate distribution in percentage and then sort by index from earliest to latest

In [14]:
#date_crawled
date_crawled_ext = autos['date_crawled'].str[:10]
date_crawled_ext.value_counts(normalize=True, dropna=False).sort_index().head()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
Name: date_crawled, dtype: float64

In [15]:
#ad_created

ad_created_ext = autos['ad_created'].str[:10]
ad_created_ext.value_counts(normalize=True,dropna=False).sort_index().head()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
Name: ad_created, dtype: float64

In [16]:
#last_seen

last_seen_ext = autos['last_seen'].str[:10]
last_seen_ext.value_counts(normalize=True,dropna=False).sort_index().tail()

2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [17]:
#Exploring car registraton year

autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The above analysis shows that some values in __registration_year__ column are inaccurate.

As we can see that the minimum value is 1000 year befor the invention of cars and maximum value of 9999 which exceed the last year car was seen online 2016

We will have to deal with those incorrect values

<b>Deciding which value to remove and retain

In [18]:
autos['registration_year'].value_counts().sort_index(ascending=True).head(20)
autos['registration_year'].value_counts().sort_index(ascending=True).tail(20)

2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

Above we use value_counts().sort_index to see those values that are out of range and we decided to retain year from 1910 to 2016 since they are the most reasonable years of registration


In [19]:
autos = autos[autos['registration_year'].between(1910, 2016)]

In [20]:
#Checking if what we did is correct
autos['registration_year'].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

We now have accurate values in our __registration_year__ column as we can see that the minimum value is 1910 and maximum value is 2016

Now we want to calculate the distribution of __registration_year__ column

In [21]:
reg_yr_dist = autos['registration_year'].value_counts(normalize=True)
reg_yr_dist.head()

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
Name: registration_year, dtype: float64

### Exploring Car Price by Brands

<b> Exploring by most popular Car Brands

In [22]:
brands = autos['brand'].value_counts(normalize=True)
brands.head()

volkswagen       0.211264
bmw              0.110045
opel             0.107581
mercedes_benz    0.096463
audi             0.086566
Name: brand, dtype: float64

The above code is used to know the distribution of car brand in percentages from which we select those greater than  .05%

In [23]:
popular_brands = brands[brands > .05].index
popular_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

The above brands are those that met the condition  stated

### Mean Price

In [24]:
#Aggregate by  brand to know calculate the mean price
avg_price_brand = {}

for brand in popular_brands:
    brand_mean = autos.loc[autos['brand'] == brand, 'price'].mean()
    avg_price_brand[brand] = int(brand_mean)
avg_price_brand

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

Out of the 6 most popular brands, there is a distinct price gap:

* Audi, Mercedes Benz and BMW are more expensive
* Ford and Opel are less expensive
* Volkswagen is in between 

In [25]:
#Convert to series object

avg_price_brand_series = pd.Series(avg_price_brand).sort_values(ascending=False)
avg_price_brand_series

audi             9336
mercedes_benz    8628
bmw              8332
volkswagen       5402
ford             3749
opel             2975
dtype: int64

In [26]:
#Convert the   series object to DataFrame

avg_price_brand_df = pd.DataFrame(avg_price_brand_series, columns=['mean_price'])
avg_price_brand_df

mean_price
audi                 9336
mercedes_benz        8628
bmw                  8332
volkswagen           5402
ford                 3749
opel                 2975

### Mean Mileage

In [27]:
avg_mileage = {}
for brand in popular_brands:
    is_mileage = autos[autos['brand'] == brand]
    mean_mileage = is_mileage['odometer_kg'].mean()
    avg_mileage[brand] = int(mean_mileage)
    
avg_mileage

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [28]:
avg_mileage_series = pd.Series(avg_mileage).sort_values(ascending=False)
avg_mileage_series

bmw              132572
mercedes_benz    130788
opel             129310
audi             129157
volkswagen       128707
ford             124266
dtype: int64

In [29]:
#Assigning mean_mileage series as a new column to the DataFrame we  created above

avg_price_brand_df['mean_mileage'] = avg_mileage_series

In [30]:
avg_price_brand_df

mean_price  mean_mileage
audi                 9336        129157
mercedes_benz        8628        130788
bmw                  8332        132572
volkswagen           5402        128707
ford                 3749        124266
opel                 2975        129310

<b> Correcting gearbox, fuel_type and unrepaired_damage column values
    
Changing the values from German to English
    

In [31]:
autos['gearbox'].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [32]:
corrections = {
    'manuell':'manual',
    'automatik':'automatic',
}
autos['gearbox'] = autos['gearbox'].map(corrections)

#check if correct
autos['gearbox'].unique()

array(['manual', 'automatic', nan], dtype=object)

In [33]:
corrections = {
    'nein':'No',
    'ja':'Yes'
}
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(corrections)

#check if correct
autos['unrepaired_damage'].unique()

array(['No', nan, 'Yes'], dtype=object)

In [34]:
corrections = {
    'lpg':'lpg',
    'benzin':'petrol',
    'diesel':'diesel',
    'cng':'cng',
    'hybrid':'hybrid',
    'elektro':'electro',
    'andere':'others'
}
autos['fuel_type'] = autos['fuel_type'].map(corrections)

#check if correct
autos['fuel_type'].unique()

array(['lpg', 'petrol', 'diesel', nan, 'cng', 'hybrid', 'electro',
       'others'], dtype=object)

In [35]:
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type  price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot   5000  control          bus               2004     manual   
1    Angebot   8500  control    limousine               1997  automatic   
2    Angebot   8990     test    limousine               2009     manual   

   power_ps   model  odometer_kg  registration_month fuel_type       brand  \
0       158  andere       150000                   3       lpg     peugeot   
1       286     7er       150000                   6    petrol         bmw   
2       102    golf        70000                   7    petrol  volkswagen   

  unrepaired_damage           ad_created  no_of_pictures  postalCode  \
0                No  2016-03-26 00:00:00               0       79588   
1                No  2016-04-04 00:00:00               0       71034   
2                No  2016-03-26 00:00:00               0       35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

#### Converting Date Columns to Numeric

Below we extract the first ten characters from te column then replace __-__ with an empty string and finally convert to numeric dtype

In [36]:
autos['date_crawled'] = autos['date_crawled'].str[:10].str.replace('-','').astype(int)
autos['ad_created'] = autos['ad_created'].str[:10].str.replace('-','').astype(int)
autos['last_seen'] = autos['last_seen'].str[:10].str.replace('-','').astype(int)

In [37]:
autos.head()

date_crawled                                               name  seller  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2      20160326                         Volkswagen_Golf_1.6_United  privat   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  privat   

  offer_type  price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot   5000  control          bus               2004     manual   
1    Angebot   8500  control    limousine               1997  automatic   
2    Angebot   8990     test    limousine               2009     manual   
3    Angebot   4350  control   kleinwagen               2007  automatic   
4    Angebot   1350     test        kombi               2003     manual   

   power_ps   model  odometer_kg  registration_month fuel_type       brand  \
0       158  andere       150000                   3       lpg     peugeot   
1       286     7er       150000                   6    petrol         bmw   
2       102    golf        70000                   7    petrol  volkswagen   
3        71  fortwo        70000                   6    petrol       smart   
4         0   focus       150000                   7    petrol        ford   

  unrepaired_damage  ad_created  no_of_pictures  postalCode  last_seen  
0                No    20160326               0       79588   20160406  
1                No    20160404               0       71034   20160406  
2                No    20160326               0       35394   20160406  
3                No    20160312               0       33729   20160315  
4                No    20160401               0       39218   20160401

In [38]:
s = autos['name'].str.split('_').str[:2]
s.tail(20)

49979    [Volkswagen, Polo]
49980        [Ford, Escort]
49981         [Opel, Astra]
49982        [Skoda, Fabia]
49983         [Ford, focus]
49985    [Verkaufe, meinen]
49986      [Chrysler, 300C]
49987            [Audi, A3]
49988            [BMW, 330]
49989            [VW, Polo]
49990      [Mercedes, Benz]
49991          [Kleinwagen]
49992        [Fiat, Grande]
49993            [Audi, A3]
49994            [Audi, A6]
49995            [Audi, Q5]
49996         [Opel, Astra]
49997           [Fiat, 500]
49998            [Audi, A3]
49999        [Opel, Vectra]
Name: name, dtype: object

### Analyzing Brand/Model Combinations

What are the most common brand/model combinations ?

In [39]:
brands_models = autos.loc[autos['brand'].isin(popular_brands)].groupby(['brand', 'model']).size().sort_values(ascending=False)
brands_models.head(10)

brand          model   
volkswagen     golf        3707
bmw            3er         2615
volkswagen     polo        1609
opel           corsa       1592
volkswagen     passat      1349
opel           astra       1348
audi           a4          1231
mercedes_benz  c_klasse    1136
bmw            5er         1132
mercedes_benz  e_klasse     958
dtype: int64

Among the top 10 Models, 3 of Volkswagen model falls in the top 10, we won't be surprise as we already knew that Volkswagen is the most popular brand in this dataset. Therefore, the most popular model is Volkswagen Golf. While Bmw, Opel and Mercedes Benz, each have two models in the top ten.

In [40]:
top3_brand_models = (brands_models.reset_index() #creates a basic table
                          .sort_values(['brand',0], ascending=[True,False]) #sorting by brand and then by count
                          .set_index(['brand','model']) #this will make showing the brand only once
                          .groupby('brand').head(3) # restricting to the top 3
                    .rename(columns={0: 'count'})) 
top3_brand_models


count
brand         model          
audi          a4         1231
              a3          825
              a6          797
bmw           3er        2615
              5er        1132
              1er         521
ford          focus       762
              fiesta      722
              mondeo      479
mercedes_benz c_klasse   1136
              e_klasse    958
              a_klasse    539
opel          corsa      1592
              astra      1348
              vectra      544
volkswagen    golf       3707
              polo       1609
              passat     1349

<b> How much cheaper are cars with damage than their non-damaged counterparts?

In [41]:
price_damaged = autos.groupby(['unrepaired_damage'])['price'].mean().astype(int).reset_index().rename(columns={'price': 'avg_price'})
price_damaged

unrepaired_damage  avg_price
0                No       7164
1               Yes       2241

Above it is shown that those damaged cars are lot more cheaper than those that are not damaged. On average they are 3 times more cheaper

<b> Odometer_km Average Price

In [42]:
odometer_avg_price = autos.groupby(['odometer_kg'])['price'].mean().astype(int).reset_index().rename(columns={'price': 'avg_price'})
odometer_avg_price

odometer_kg  avg_price
0          5000       8873
1         10000      20550
2         20000      18448
3         30000      16608
4         40000      15499
5         50000      13812
6         60000      12385
7         70000      10927
8         80000       9721
9         90000       8465
10       100000       8132
11       125000       6214
12       150000       3767

The analoysis above shows that cars with odometer_km between 10,000 to 70,000 has the highest price on average

## Conclusions
In this project we have performed explorative analysis, after cleaning the data, of a subset of used cars ads from a German advertising website __eBay Kleinanzeigen__ 

Our findings are:

* The most advertised brands are Volkswagen, Bmw, Opel, Mercedes_benz, Audi and Ford.
* Volkswagen is also on average the a middle price car;
* The most popular car overall is Volkswagen Golf;
* Cars with unrepaired damages are up to 66% cheaper than those which do not have unrepaired damages;

Further analysis might include:

* Exploring the correlation between the age of the car and it's price;
* Checking which brand sells faster (e.g. by looking at date of ad creating and last seen date);
* Finding correlations between prices and locations (according to postcode), in order to see if cars are more expensive in certain areas.